### ISSUE TITLE
Writing past 2GB corrupts disk image

### ISSUE THREAD

**Issue by FelicitusNeko** [joncampbell123/dosbox-x#1073]
**Describe the bug**
Any attempts to read/write on a mounted disk image past 2 GB will result in failure. Any attempt to write past 2 GB will cause the disk image to become corrupted.

**To Reproduce**
There are many ways to get to this, but here's the shortest I can think of:
1. Create disk image with built-in `IMGMAKE` utility, any size larger than 2 GB
2. Mount the image, as well as an MS-DOS boot disk
3. `boot -l a`
4. Use `fdisk` to create a partition
5. Restart DOSBox
6. Mount boot disk and image
7. `boot -l a`
8. Use ScanDisk to perform a surface scan
9. "ScanDisk cannot read from the last cluster on drive C."
10. Choose to continue
11. Wait for it to get to 2 GB
12. Unreadable cluster error for the remainder of the disk

**Expected behavior**
ScanDisk does not report unreadable last sector, and is able to check the image past 2 GB. Other write operations past 2 GB do not cause image corruption.

**Screenshots**
![image](https://user-images.githubusercontent.com/50271878/57182111-e8f15280-6e71-11e9-9a72-e9e6e649c07f.png)

**Environment (please complete the following information):**
 - Windows 10 Pro 1809
 - DOSBox-X 0.82.18 (EmuCR build)
 - Configuration: Mix between default settings and conf provided by http://dosbox95.darktraveler.com/ (as my ultimate goal is to install Windows 98)
[dosbox-take4.conf.txt](https://github.com/joncampbell123/dosbox-x/files/3144579/dosbox-take4.conf.txt)

**Additional context**
I have tried a few different setups: creating the image with IMGMAKE or WinImage; using an MS-DOS boot disk or a FreeDOS boot disk; using an external tool to force the partition into LBA mode; nothing seems to work.

**Comment by nckstwrt** [comment]
So, I hit this exact same issue today. Could not find out why Scandisk was b!tching on any drive bigger than 2gb (running Dosbox-X on Windows10)

From a quick debug, I think the issue is at:
Bit8u imageDisk::Read_AbsoluteSector(Bit32u sectnum, void * data)
in bios_disk.cpp at this code:
fseeko64(diskimg,(long)bytenum,SEEK_SET);
    res = (Bit64u)ftello64(diskimg);
    if (res != bytenum) {
        LOG_MSG("fseek() failed in Read_AbsoluteSector for sector %lu. Want=%llu Got=%llu\n",
            (unsigned long)sectnum,(unsigned long long)bytenum,(unsigned long long)res);
        return 0x05;
    }

^ The trouble is the cast to long in the fseek (why are we doing that?). remove the (long) and all works well.

**Comment by Allofich** [comment]
The cast to `long` was added here:
https://github.com/joncampbell123/dosbox-x/commit/b86130bd4547aadf0a24e9587d8f5a07e399f112

Looks like it was previously a cast to `off_t`.

**Comment by joncampbell123** [comment]
Visual Studio is the only compiler supported to use __int64 or long for fseek() instead of off_t

**Comment by nckstwrt** [comment]
off_t is just long (32 bit) in Visual Studio I think (so that cast would be bad in this case too). Does compilation break elsewhere if we do not cast at all?

**Comment by joncampbell123** [comment]
Please try the latest commit, which defines it's own "off_t" to satisfy Linux's need for off_t and Visual Studio's need for __int64 as data types for file seeking.

**Comment by joncampbell123** [comment]
Can you compile from scratch?

**Comment by nckstwrt** [comment]
Only just got chance to try it all from scratch (compiling in VS2019) - works great with a 4gb hd img file - no issues scandisk'ing, etc

Many thanks!

**Comment by KGOrphanides** [comment]
This bug continues to affect 4GB disk images on Raspberry Pi 4 (4GB model, running Rasbian Buster, kernel version 4.19.118-v71+). As expected, it doesn't affect 2GB images.

It also does not occur on my regular desktop (an x86-64 system running POP!_OS 20.04 focal).

**Comment by AlisterAmo** [comment]
at the time of this writing, using latest raspberry OS and latest version of dosbox-x code in this repo (compiled yesterday), the problem persist in the same way @KGOrphanides explained more than a year ago.
If someone is so kind to explain what part of the code should be changed in order to perform the type casting that causes this issue in rpi (and all linuxes maybe?), I would sincerely appreciate.

**Comment by joncampbell123** [comment]
Raspian is still 32-bit, right?

DOSBox-X should be using lseek64, if it is not, then lseek() will wrap around at 2GB and cause problems.

CXXFLAGS should include -D_FILE_OFFSET_BITS=64 on 32-bit compiles so the C++ headers map lseek to lseek64.

**Comment by joncampbell123** [comment]
Another possibility is the code leading up to lseek() is using int or long, which are 32-bit on 32-bit systems. 64-bit Linux sets int to 32-bit and long to 64-bit. That code should be using long long instead to get 64-bit offsets on 32-bit and 64-bit builds.

**Comment by AlisterAmo** [comment]
hi @joncampbell123.
Yes, raspberry pi os is still 32bit.
I will tray a build with the aforementioned flags.
But to me, it sounds that a change in code so it uses long long should be the permanent solution, right?

**Comment by joncampbell123** [comment]
@AlisterAmo Right. Maybe int64_t if anyone is still compiling with older versions of Microsoft Visual C++ that do not support "long long".

**Comment by AlisterAmo** [comment]
Probably I have mode some mistake, but the new build I have compiled does not change the behavior.
I have performed:
`CXXFLAGS=-D_FILE_OFFSET_BITS=64 ./build`

**Comment by sekushi18** [comment]
Sorry if it's not helping. but has this been fixed yet? I'm having issues as my imgmake will work but as it is mounted it causes the .img to only display corrupted name and 1 or 2 files that's corrupted. saying 100% of file is full. running scan disk just causes it the unmount the drive and throw a can't read error.
